### PR TITLE
Enable validation of backend certificates in BIG-IP server SSL profiles

### PR DESCRIPTION
**Pull request by f5gary**
Problem:
Server SSL profiles do not validate the backend certificates by default
when the BIG-IP is configured with one or more reencrypt OpenShift
Routes.

Solution:
Made the following changes to the controller related to server-side SSL
profiles.
1. Added a new annotation "virtual-server.f5.com/secure-serverssl"
   which if set to true will enable certificate validation on the
   backend certificate. The default is false, which ignores the
   validation as before.
2. Moved creation of the default server-side SSL profile to the startup
   code and added the proper checks to ensure it is always configured.
3. CA files are represented in the controler as a custom profile with
   the new CAFile member set to "self". These profiles exist only to create
   the correct certificate on the BIG-IP and do not result in a new SSL
   profile.
4. The custom profile struct has also been updated with a new PeerCertMode
   member which determines if the profile should validate the backend
   certificate (if it is set to "require") which causes the BIG-IP to use
   CAFile to validate the backend certificate. If PeerCertMode is "ignore"
   or blank (the default) then the backend certificate is not validated.
   This is backward compatible with previous versions of the controller.
5. Since the virtual servers that are maintained for OpenShift Route
   objects can contain multiple server-side SSL profiles and needs help to
   know which one to use for different backends, the existing iRule that
   is used currently for passthrough and reencrypt routes has been
   modified so it now handles SERVER_CONNECTED events and does a lookup
   in a new internal data group that maps backend server names to SSL
   profiles and sets the correct profile on that connection accordingly.
6. Reworked the internal data structure for data groups so there is one
   map per namespace, which is required by the requirements of the new SSL
   profile data group. This also helps clean up existing data group
   maintenance related to namespaces.

Also created some constants for annotations where they were missing,
and some other miscellaneous cleanup.

affects-branches: master

### CODE REVIEW

File: pkg/appmanager/outputConfig.go [hunk 1]
```diff
@@ -154,6 +161,8 @@ func (appMgr *Manager) outputConfigLocked() {
 				}
 				log.Infof("Wrote %v Virtual Server and %v IApp configs",
 					virtualCount, iappCount)
+				grout, _ := json.MarshalIndent(resources, "", "  ")
+				log.Infof("%s", string(grout))
```
[the remark this answers is not in the document]
Comment (f5gary) [reply pkg/appmanager/outputConfig.go]: No, that should have been removed. I will do that.

File: docs/README.rst [hunk 2]
```diff
@@ -600,31 +600,34 @@ Supported Route Configurations
...
+|                      | interval        | integer     | Required  | The interval at which to check the health of the virtual server.               | N/A         |
+|                      |                 |             | [#hm1]_   |                                                                                |             |
++----------------------+-----------------+-------------+-----------+--------------------------------------------------------------------------------+-------------+
+|                      | timeout         | integer     | Required  | Number of seconds before the check times out.                                  | N/A         |
+|                      |                 |             | [#hm1]_   |                                                                                |             |
++----------------------+-----------------+-------------+-----------+--------------------------------------------------------------------------------+-------------+
+| virtual-server.f5.com/secure-serverssl | boolean     | Optional  | Specifies to validate the server-side SSL certificate of Re-encrypt            | false       |
```
[the remark this answers is not in the document]
Comment (f5gary) [reply docs/README.rst]: Will do.

File: python/bigipconfigdriver.py [hunk 3]
```diff
@@ -229,13 +229,25 @@ def _create_custom_profiles(mgmt, partition, custom_profiles):
...
+    # Server profiles may reference a CA cert in another server profile.
+    # These need to be loaded first.
+    for profile in custom_profiles:
+        caFile = profile.get('caFile', '')
+        if profile['context'] == 'serverside' and caFile == "self":
+            incomplete += create_server_ssl_profile(mgmt, partition, profile)
+            customProfiles = True
+
     for profile in custom_profiles:
         if profile['context'] == 'clientside':
```
[the remark this answers is not in the document]
Comment (f5gary) [reply python/bigipconfigdriver.py]: The profiles with caFile == 'self' need to be handled before any other profile that references them or creation of the referencing profile will fail. 2 loops should be better than sorting them first which is why it is this way.

[the remark this answers is not in the document]
Comment (sjberman) [reply python/bigipconfigdriver.py]: Ok, that makes sense.

**Pull request merged**